### PR TITLE
Add  watchdog functionalities

### PR DESCRIPTION
**Pull request by ya0guang**
## Description

Clients can now cancel a task, even when the task is executing;
Teaclave can detect `dangling tasks`, which means the executor can send heartbeat packets for liveness report.

Currently Client Rust API changes are missing.

Fixes # (issue)

## Type of change (select or add applied and delete the others)

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] API change with a documentation update
- [X] Additional test coverage
- [ ] Code cleanup or just sync with upstream third-party crates

## Checklist

- [X] Fork the repo and create your branch from `master`.
- [X] If you've added code that should be tested, add tests.
- [ ] If you've changed APIs, update the documentation.
- [ ] Ensure the tests pass (see CI results).
- [ ] Make sure your code lints/format.

### CODE REVIEW

File: services/execution/enclave/src/service.rs [hunk 1]
```diff
@@ -61,47 +66,95 @@ impl TeaclaveExecutionService {
...
             scheduler_client,
             fusion_base: fusion_base.as_ref().to_owned(),
+            id: Uuid::new_v4(),
+            status: ExecutorStatus::Idle,
         })
     }
 
     pub(crate) fn start(&mut self) -> Result<()> {
+        let (tx, rx) = mpsc::channel();
+        let mut current_task = Arc::new(None);
```
Comment (ya0guang) [inline services/execution/enclave/src/service.rs:76]: type annot

**Review by bigdata-memory — COMMENTED** [review]
Some of commits in this PR looks like intermediate patches, so it would be better if squash some of them with more meaningful commit comments, Thanks.

File: services/execution/enclave/src/lib.rs [hunk 2]
```diff
@@ -77,18 +77,17 @@ fn start_service(config: &RuntimeConfig) -> Result<()> {
...
 
 #[handle_ecall]
 fn handle_start_service(input: &StartServiceInput) -> TeeServiceResult<StartServiceOutput> {
     match start_service(&input.config) {
         Ok(_) => Ok(StartServiceOutput),
+        // terminate the enclave for executor
         Err(e) => {
-            log::error!("Failed to start the service: {}", e);
-            Err(TeeServiceError::ServiceError)
+            log::error!("Service Shut Donw, Reason: {}", e);
```
Comment (mssun) [inline services/execution/enclave/src/lib.rs:89]: `Service shutdown, reason: {}`

File: examples/python/builtin_gbdt_train.py [hunk 3]
```diff
@@ -103,6 +103,13 @@ def gbdt(self):
         print("[+] invoking task")
         client.invoke_task(task_id)
 
+        # If you feel like you don't want to wait for the task to finish
+        # or something unexpected happens,
+        # you may uncomment the following lines to cancel the task
+        # time.sleep(3)
+        # print("[+] canceling task")
```
[the remark this answers is not in the document]
Comment (ya0guang) [reply examples/python/builtin_gbdt_train.py:110]: The code can be uncomment to cancel an executing task. It means to demo calling the new `cancel_task` API and provide users an example. This example will run for a long while, so users may want to know how to cancel a task running for a long period without response.

File: cmake/scripts/test.sh [hunk 4]
```diff
@@ -97,7 +97,7 @@ run_functional_tests() {
   pushd ${TEACLAVE_SERVICE_INSTALL_DIR}
   ./teaclave_authentication_service &
   ./teaclave_storage_service &
-  sleep 3    # wait for authentication and storage service
+  sleep 10    # wait for authentication and storage service
```
[the remark this answers is not in the document]
Comment (ya0guang) [reply cmake/scripts/test.sh:100]: Agree

[the remark this answers is not in the document]
Comment (ya0guang) [reply cmake/scripts/test.sh:100]: I think probing may be added in the following commits

**Review by mssun — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**